File: lib/providers/library_provider.dart
import 'dart:convert';
import 'package:flutter/material.dart';
import 'package:uuid/uuid.dart';
import '../models/saved_project.dart';
import '../models/snippet.dart';
import '../services/preferences_service.dart';

class LibraryProvider with ChangeNotifier {
  final PreferencesService _prefsService = PreferencesService();
  List<SavedProject> _projects = [];
  List<Snippet> _snippets = [];

  List<SavedProject> get projects => _projects;
  List<Snippet> get snippets => _snippets;

  LibraryProvider() {
    _loadLibrary();
  }

  Future<void> _loadLibrary() async {
    final projectsJson = await _prefsService.loadStringList(PreferencesService.keySavedProjects);
    if (projectsJson != null) {
      _projects = projectsJson
          .map((str) => SavedProject.fromJson(jsonDecode(str)))
          .toList();
    }

    final snippetsJson = await _prefsService.loadStringList(PreferencesService.keySavedSnippets);
    if (snippetsJson != null) {
      _snippets = snippetsJson
          .map((str) => Snippet.fromJson(jsonDecode(str)))
          .toList();
    }

    notifyListeners();
  }

  Future<void> _saveLibrary() async {
    final projectsJson = _projects.map((p) => jsonEncode(p.toJson())).toList();
    await _prefsService.saveStringList(PreferencesService.keySavedProjects, projectsJson);

    final snippetsJson = _snippets.map((s) => jsonEncode(s.toJson())).toList();
    await _prefsService.saveStringList(PreferencesService.keySavedSnippets, snippetsJson);
  }

  void saveProject({
    required String name,
    required String description,
    required List<String> tags,
    required String jsonContent,
    String? id,
  }) {
    final now = DateTime.now();
    if (id != null) {
      // Update existing
      final index = _projects.indexWhere((p) => p.id == id);
      if (index != -1) {
        _projects[index] = SavedProject(
          id: id,
          name: name,
          description: description,
          tags: tags,
          lastModified: now,
          jsonContent: jsonContent,
        );
      }
    } else {
      // Create new
      final newProject = SavedProject(
        id: const Uuid().v4(),
        name: name,
        description: description,
        tags: tags,
        lastModified: now,
        jsonContent: jsonContent,
      );
      _projects.add(newProject);
    }
    _saveLibrary();
    notifyListeners();
  }

  void deleteProject(String id) {
    _projects.removeWhere((p) => p.id == id);
    _saveLibrary();
    notifyListeners();
  }

  void saveSnippet({required String name, required String elementJson}) {
    final snippet = Snippet(
      id: const Uuid().v4(),
      name: name,
      elementJson: elementJson,
    );
    _snippets.add(snippet);
    _saveLibrary();
    notifyListeners();
  }

  void deleteSnippet(String id) {
    _snippets.removeWhere((s) => s.id == id);
    _saveLibrary();
    notifyListeners();
  }
}

File: src/providers/LibraryProvider.js
// LibraryProvider: Manages library state (converted from Flutter LibraryProvider)
import React, { createContext, useContext, useEffect, useState } from 'react';
import { v4 as uuidv4 } from 'uuid';
import { PreferenceKeys, PreferencesService } from '../services/preferencesService';
import { SavedProject } from '../models/SavedProject';
import { Snippet } from '../models/Snippet';

export const LibraryContext = createContext();

export const LibraryProvider = ({ children }) => {
  const [projects, setProjects] = useState([]);
  const [snippets, setSnippets] = useState([]);
  const [isFirebaseAvailable, setIsFirebaseAvailable] = useState(true);

  useEffect(() => {
    loadLibrary();
  }, []);

  const loadLibrary = async () => {
    try {
      const savedProjects = await PreferencesService.getJson(PreferenceKeys.savedProjects, []);
      const savedSnippets = await PreferencesService.getJson(PreferenceKeys.savedSnippets, []);
      if (Array.isArray(savedProjects)) {
        setProjects(savedProjects.map((p) => SavedProject.fromJson(p)));
      }
      if (Array.isArray(savedSnippets)) {
        setSnippets(savedSnippets.map((s) => Snippet.fromJson(s)));
      }
    } catch (error) {
      console.error('Failed to load library', error);
    }
  };

  const persistProjects = async (nextProjects) => {
    setProjects(nextProjects);
    await PreferencesService.setJson(
      PreferenceKeys.savedProjects,
      nextProjects.map((project) => project.toJson())
    );
  };

  const persistSnippets = async (nextSnippets) => {
    setSnippets(nextSnippets);
    await PreferencesService.setJson(
      PreferenceKeys.savedSnippets,
      nextSnippets.map((snippet) => snippet.toJson())
    );
  };

  const saveProject = ({ name, description = '', tags = [], jsonContent }) => {
    const project = new SavedProject({
      id: uuidv4(),
      name,
      description,
      tags,
      jsonContent,
      lastModified: new Date().toISOString(),
    });
    const next = [project, ...projects];
    persistProjects(next);
    return project;
  };

  const updateProject = (id, patch = {}) => {
    const next = projects.map((project) =>
      project.id === id
        ? new SavedProject({
            ...project.toJson(),
            ...patch,
            lastModified: new Date().toISOString(),
          })
        : project
    );
    persistProjects(next);
  };

  const deleteProject = (id) => {
    const next = projects.filter((project) => project.id !== id);
    persistProjects(next);
  };

  const saveSnippet = ({ name, elementJson }) => {
    const snippet = new Snippet({ id: uuidv4(), name, elementJson });
    const next = [snippet, ...snippets];
    persistSnippets(next);
    return snippet;
  };

  const deleteSnippet = (id) => {
    const next = snippets.filter((snippet) => snippet.id !== id);
    persistSnippets(next);
  };

  return (
    <LibraryContext.Provider
      value={{
        projects,
        snippets,
        isFirebaseAvailable,
        setIsFirebaseAvailable,
        saveProject,
        updateProject,
        deleteProject,
        saveSnippet,
        deleteSnippet,
      }}
    >
      {children}
    </LibraryContext.Provider>
  );
};

export const useLibrary = () => useContext(LibraryContext);
